pageserver/src/tenant/timeline/import_pgdata.rs

```diff
@@ -33,166 +44,214 @@ pub async fn doit(
+                        Lsn(0),   // no ancestor lsn
+                        base_lsn, // latest_gc_cutoff_lsn
+                        base_lsn, // initdb_lsn
+                        pg_version,
+                    );
 
-                let _start_lsn = disk_consistent_lsn + 1;
+                    let _start_lsn = disk_consistent_lsn + 1;
 
-                timeline
-                    .remote_client
-                    .schedule_index_upload_for_full_metadata_update(&metadata)?;
+                    timeline
+                        .remote_client
+                        .schedule_index_upload_for_full_metadata_update(&metadata)?;
 
-                timeline.remote_client.wait_completion().await?;
+                    timeline.remote_client.wait_completion().await?;
 
-                anyhow::Ok(())
+                    anyhow::Ok(())
+                }
             }
+            .await?;
+
+            flow::run(timeline.clone(), control_file, storage.clone(), ctx).await?;
+
+            //
+            // Communicate that shard is done.
+            // Ensure at-least-once delivery of the upcall to storage controller
+            // before we mark the task as done and never come here again.
+            //
+            storcon_client
+                .put_timeline_import_status(
+                    timeline.tenant_shard_id,
+                    timeline.timeline_id,
+                    // TODO(vlad): What about import errors?
+                    ShardImportStatus::Finalizing,
+                )
+                .await
+                .map_err(|_err| {
+                    anyhow::anyhow!("Shut down while putting timeline import status")
+                })?;
+
+            //
+            // Mark as done in index_part.
+            // This makes subsequent timeline loads enter the normal load code path
+            // instead of spawning the import task and calling this here function.
+            //
+            info!("mark import as complete in index part");
+            timeline
+                .remote_client
+                .schedule_index_upload_for_import_pgdata_state_update(Some(
+                    index_part_format::Root::V1(index_part_format::V1::Done(
+                        index_part_format::Done {
+                            idempotency_key,
+                            started_at,
+                            finished_at: chrono::Utc::now().naive_utc(),
+                        },
+                    )),
+                ))?;
+
+            timeline.remote_client.wait_completion().await?;
+
+            storcon_client
+                .put_timeline_import_status(
+                    timeline.tenant_shard_id,
+                    timeline.timeline_id,
+                    ShardImportStatus::Done,
+                )
+                .await
+                .map_err(|_err| {
+                    anyhow::anyhow!("Shut down while putting timeline import status")
+                })?;
```

Comment:
Assuming the index upload goes through but then we crash.

Will we ever upcall into storcon to noti;fy about import status done?

Reply:
I ended up moving the final index part update into a PS route handler that storcon calls to finalize the import.
I tried quite hard to wiggle my way out of doing this, but didn't find a better solution: https://github.com/neondatabase/neon/pull/11855/commits/e50f8e9815ffb2bbb3dc91c35b4ac84ecb7fd944